examples/audio/melgan_spectrogram_inversion.py

```diff
@@ -0,0 +1,619 @@
+below
+"""
+
+# Custom keras layer for on-the-fly audio to spectrogram conversion
+
+
+class MelSpec(tf.keras.layers.Layer):
+    def __init__(
+        self,
+        frame_length=1024,
+        frame_step=256,
+        fft_length=None,
+        sampling_rate=22050,
+        n_mel_channels=80,
+        fmin=125,
+        fmax=7600,
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+        self.frame_length = frame_length
+        self.frame_step = frame_step
+        self.fft_length = fft_length
+        self.sampling_rate = sampling_rate
+        self.n_mel_channels = n_mel_channels
+        self.fmin = fmin
+        self.fmax = fmax
+        # Defining mel filter. This filter will be multiplied with the STFT output
+        self.mel_filterbank = tf.signal.linear_to_mel_weight_matrix(
+            num_mel_bins=self.n_mel_channels,
+            num_spectrogram_bins=self.frame_length // 2 + 1,
+            sample_rate=self.sampling_rate,
+            lower_edge_hertz=self.fmin,
+            upper_edge_hertz=self.fmax,
+        )
+
+    def call(self, audio, training=True):
+        # We will only perform the transformation during training.
+        if training:
+            # Taking the Short Time Fourier Transform. Ensure that the audio is padded.
+            # In the paper, the STFT output is padded using the 'REFLECT' strategy.
+            stft = tf.signal.stft(
+                tf.squeeze(audio, -1),
+                self.frame_length,
+                self.frame_step,
+                self.fft_length,
+                pad_end=True,
+            )
+
+            # Taking the magnitude of the STFT output
+            magnitude = tf.abs(stft)
+
+# Multiplying the Mel-filterbank with the magnitude and scaling it using the db scale
+            mel = tf.matmul(tf.square(magnitude), self.mel_filterbank)
+            log_mel_spec = tfio.audio.dbscale(mel, 80)
+            return log_mel_spec
+        else:
+            return audio
+
+    def get_config(self):
+        config = super(MelSpec, self).get_config()
+        config.update(
+            {
+                "frame_length": self.frame_length,
+                "frame_step": self.frame_step,
+                "fft_length": self.fft_length,
+                "sampling_rate": self.sampling_rate,
+                "n_mel_channels": self.n_mel_channels,
+                "fmin": self.fmin,
+                "fmax": self.fmax,
+            }
+        )
+        return config
+
+
+"""
+The Residual Convolutional Block extensively uses dilations and has a total receptive
+field of 27 timesteps per block. The dilations must grow as a power of the `kernel_size`
+to ensure reduction of hissing noise in the output. The network proposed by the paper is
+as follows:
+
+<p align="center">
+<img src="https://i.imgur.com/gENKH91.png" alt="drawing" height="500"/>
+</p>
+"""
+
+# Creating the Residual Stack Layer
+
+
+class ResidualStack(tf.keras.layers.Layer):
```

Comment:
Make this a Functional model building function (see guide https://github.com/keras-team/keras-io/blob/master/contributor_guide.md)

You will save a lot of space that way (and make the code more readable)

Reply:
I wanted to discuss this. The reason why I didn't implement it as a functional model in the first place was that the WeightNormalization layer from tf-addons does not support a dynamic channel size and there wasn't a way in which I could fix the `input_shape` for the `Input` layer in the functional model ( As you already know, `keras.Model` cannot take inputs directly from external/previous layers. It has to come through an `Input` layer). I have an alternative implementation that is not exactly using the `keras.Model` API but it is purely function-based and provides decent readability. I have changed the three custom layers to this function-based implementation but I want your review on those since I am not completely satisfied with this kind of structuring. We will go ahead with whichever method you prefer.